<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2027-1</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Rodrigo Vega Vilchis
### Joel Gómez Uribe
---

<h1 align="center">Programación para la física computacional</h1> 

## Práctica 2.  Fundamentos de programacion para la fisica computacional
---
<h2>Diaz Castro Gerson Joshua </h2> 
---

### **EJERCICIOS**:

$\;$

1. La constante de Madelung

En física de la materia condensada, la constante de Madelung da el potencial eléctrico total que siente un átomo en un sólido; y depende de las cargas de los otros átomos cercanos y de sus ubicaciones. Por ejemplo, el cristal de cloruro de sodio sólido (la sal de mesa), tiene átomos dispuestos en una red cúbica, con átomos de sodio y cloro alternados, teniendo los de sodio una carga positiva $+e$ y los de cloro una negativa $-e$, (donde $e$ es la carga del electrón). Si etiquetamos cada posición en la red con tres coordenadas enteras $(i,j,k)$, entonces los átomos de sodio caen en posiciones donde $i+j+k$ es par, y los átomos de cloro en posiciones donde $i+j+k$ es impar.

Consideremos un átomo de sodio en el origen, i.e. $i = j = k = 0$, y calculemos la constante de Madelung. Si el espaciado de los átomos en la red es $a$, entonces la distancia desde el origen al átomo en la posición $(i,j,k)$ es:

$$ \sqrt{(ia)^2 + (ja)^2 + (ka)^2} = a\sqrt{i^2 + j^2 + k^2}$$

y el potencial en el origen creado por tal átomo es:

$$V(i,j,k) = \pm \frac{e}{4\pi\epsilon_0 a \sqrt{i^2 + j^2 + k^2}}$$

siendo $\epsilon_0$ la permitividad del vacío y el signo de la expresión se toma dependiendo de si $i+j+k$ es par o impar. Así entonces, el potencial total que siente el átomo de sodio es la suma de esta cantidad sobre todos los demás átomos. Supongamos una caja cúbica alrededor del átomo de sodio en el origen, con $L$ átomos en todas las direcciones, entonces:

$$V_{\text{total}} = \sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L} V(i,j,k) = \frac{e}{4\pi\epsilon_0 a} M$$

donde $M$ es la constante de Madelung (al menos aproximadamente).

Técnicamente, la constante de Madelung es el valor de $M$ cuando $L \to \infty$, pero se puede obtener una buena aproximación simplemente usando un valor grande de $L$. Escribe un programa para calcular e imprimir la constante de Madelung para el cloruro de sodio. Utiliza un valor de $L$ tan grande como puedas, sin dejar que tu programa se ejecute en un tiempo razonable (un minuto o menos).

Solución:
Primero definimos las variables que nunca van a cambiar:

In [3]:
import numpy as np
e = 1.602e-19 #Coulombs
E_0 = 8.854e-12 #F/m
a = 2.82e-10 #m (Valor buscado en internet)
pi = np.pi

Ahora sacamos las constantes de la serie: $$V_{\text{total}} = \frac{e}{4\pi \epsilon_{0} a} \sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L}$$

De modo que solo nos quedaria resolver: $$M = \sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L} \frac{\pm 1}{\sqrt{i^2 + j^2 + k^2}}$$

No era necesario establecer las constantes jeje

In [6]:
#Definimos el valor de i,j,k,

L = 250  #El valor de L para la iteracion


Ahora programemos la suma:

In [7]:

M = 0 #La variable que ira guardando los valores
for i in range(-L,L+1):
    for j in range(-L,L+1):
        for k in range(-L,L+1):
            if i == 0 and j == 0 and k == 0:#Este es para evitar que todos los indices se hagan 0 y se indetermine
                continue
            #Calculamos la norma de nuestros indices
            norm = np.sqrt((i)**(2) + (j)**(2) + (k)**(2))
            #calculamos la suma de los indices
            suma = i + j + k
            #analizamos si es par o impar y con base a eso actualizamos a m
            p = suma%2
            if p == 0:
                M += 1/(norm) #Si es par se suma
            else:
                M -= 1/(norm) #Si no es par se resta
            
        
print(f"La constante de Madelung es aproximadamente: {M}")


La constante de Madelung es aproximadamente: -1.7452598031439428


**2. La fórmula semiempírica de la masa (FSM)** En física nuclear, la **fórmula de Weizsäcker** (conocida también como fórmula semiempírica) sirve para evaluar la masa y otras propiedades de un núcleo atómico; y está basada parcialmente en mediciones empíricas. En particular la fórmula se usa para calcular la *energía de enlace nuclear aproximada* $B$, de un núcleo atómico con número atómico $Z$ y número de masa $A$:

$$B = a_1 A - a_2 A^{2/3} - a_3 \frac{Z^2}{A^{1/3}} - a_4 \frac{(A-2Z)^2}{A} + \frac{a_5}{A^{1/2}},$$

donde, en unidades de millones de electrón-volts, las constantes son $a_1 = 15.8, a_2 = 18.3, a_3 = 0.714, a_4 = 23.2$ y

$$
a_5 = \begin{cases} 
0 & \text{si } A \text{ es impar,} \\ 
12.0 & \text{si } A \text{ y } Z \text{ son pares (ambos),} \\ 
-12.0 & \text{si } A \text{ es par y } Z \text{ impar.} 
\end{cases}
$$


a) Escribe un programa que tome como entrada los valores de $A$ y $Z$, e imprima la energía de enlace $B$ para el átomo correspondiente. Usa tu programa para encontrar la energía de enlace de un átomo con $A = 58$ y $Z = 28$ (Hint: La respuesta correcta es alrededor de los 490 MeV).

In [3]:
#Definimos las constantes
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2
A = float(input("Coloca el valor de A: "))
Z = float(input("Coloca el valor de Z: "))



Coloca el valor de A: 58
Coloca el valor de Z: 28


In [4]:
#Calculamos el valor de a5, para ello calculamos el modulo de A y de Z
modA = A%2
modZ = A%2

#Colocamos las condiciones para analizar la paridad de A y de Z, y asi poder evaluar el valor de a5

if modA != 0:
    a5 = 0
elif modA == 0 and modZ == 0:
    a5 = 12.0
elif modA == 0 and modZ != 0:
    a5 = -12.0

    #Calculamos a B

B = (a1*A) - (a2*(A**(2/3))) - (a3*((Z**2)/(A**(1/3)))) - (a4*(((A-(2*Z))**(2))/(A))) + (a5/(A**(1/2)))
print(f"El valor de de la energia de enlace 'B' es: {B}")

El valor de de la energia de enlace 'B' es: 497.5620206224374


b) Modifica el programa del inciso anterior, para escribir una segunda versión que imprima no la energía de enlace total $B$, sino la energía de unión por nucleón, que es $B / A$.

In [5]:
#Definimos las constantes
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2
A = float(input("Coloca el valor de A: "))
Z = float(input("Coloca el valor de Z: "))

#Calculamos el valor de a5, para ello calculamos el modulo de A y de Z
modA = A%2
modZ = Z%2

#Colocamos las condiciones para analizar la paridad de A y de Z, y asi poder evaluar el valor de a5

if modA != 0:
    a5 = 0
elif modA == 0 and modZ == 0:
    a5 = 12.0
elif modA == 0 and modZ != 0:
    a5 = -12.0

    #Calculamos a B/A

B = (a1*A) - (a2*(A**(2/3))) - (a3*((Z**2)/(A**(1/3)))) - (a4*(((A-(2*Z))**(2))/(A))) + (a5/(A**(1/2)))
BA = B/A
print(f"El valor de de la energia de union es: {BA}")


Coloca el valor de A: 58
Coloca el valor de Z: 28
El valor de de la energia de union es: 8.578655527973059


c) Escribe una tercera versión del programa para que tome como entrada solo un valor del número atómico $Z$ y luego pase por todos los valores de $A$ desde $A = Z$ hasta $A = 3Z$, para encontrar el que tiene la mayor energía de enlace por nucleón. Este es el núcleo más estable con el número atómico dado. Haz que tu programa imprima el valor de $A$ para este núcleo más estable y el valor de la energía de enlace por nucleón.

In [35]:
import numpy as np

In [49]:
#Definimos las constantes
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2
Z = int(input("Coloca el valor de Z: "))
valoresB = np.array([]) #Creamos un array vacio para ir guardando los valores de B
valoresBA = np.array([]) #Creamos un array vacio para ir guardando los valores de B/A


    
    



Coloca el valor de Z: 28


In [50]:
#Ahora hacemos un array con los valores de z hasta 3z
RangoA = np.arange(Z,(3*Z)+1,1)
RangoA

array([28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44,
       45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61,
       62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78,
       79, 80, 81, 82, 83, 84])

In [51]:
J = len(RangoA)
J

57

In [52]:
#Con ayuda de un ciclo for operaremos cada elemento de RangoA

for i in range(len(RangoA)):
    v = RangoA[i] #leemos el valor i-esimo del array
    #Calculamos el valor de a5, para ello calculamos el modulo de A (que en este caso seria v) y de Z
    modv = v%2
    modZ = Z%2

    #Colocamos las condiciones para analizar la paridad de A y de Z, y asi poder evaluar el valor de a5

    if modv != 0:
        a5 = 0
    elif modv == 0 and modZ == 0:
        a5 = 12.0
    elif modv == 0 and modZ != 0:
        a5 = -12.0
    #En vase a eso calculamos B Y B/A
    B = (a1*v) - (a2*(v**(2/3))) - (a3*((Z**2)/(v**(1/3)))) - (a4*(((v-(2*Z))**(2))/(v))) + (a5/(v**(1/2)))
    BA = B/v
    #Actualizamos las array con estas variables
    valoresB = np.append(valoresB,B)
    valoresBA = np.append(valoresBA,BA)

    #Ahora hay que conocer el valor maximo
maximo = valoresB.max()
#Para hayar el valor de B/A de este valor extraemos el valor correspondiente a su posicion en el array de BA

pos = list(valoresB).index(maximo)

#Extraemos el dato de valoresBA y de RangoA
BA = valoresBA[pos]
Amax = RangoA[pos]
#Imprimos nuestros valores
    
print(f"El valor de 'A' que maximiza 'B' es: {Amax}")
print(f"El valor de la energia de enlace es: {BA}")

El valor de 'A' que maximiza 'B' es: 84
El valor de la energia de enlace es: 7.537640039955393


### **3. Relatividad especial:** 
#### Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia. 

#### Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino

#### **(a)** en el marco de reposo de un observador en la Tierra y


En el marco terrestre, la distancia a recorrer es $x$ y el tiempo de viaje $t$ a velocidad constante está dado por: $$t = \frac{x}{v}$$

Sustituyendo nuestros datos:$$t = \frac{10*c años}{0.99*c}$$

In [17]:
#Entonces el programa seria:
while (True):
    v = float(input("Ingrese la velocidad como fracción de c (valor entre 0 y 1): "))
    x = float(input("Ingrese en años luz la distancia: "))
    if ((0 < v < 1) and (x > 0)):
        t = x/v
        print(f"El tiempo que tarda desde el marco de refrecia de la tierra es: {t} años")
        break
    else:
        print("Valores no validos, intente de nuevo:")

Ingrese la velocidad como fracción de c (valor entre 0 y 1): 4
Ingrese en años luz la distancia: 10
Valores no validos, intente de nuevo:
Ingrese la velocidad como fracción de c (valor entre 0 y 1): 0.99
Ingrese en años luz la distancia: 10
El tiempo que tarda desde el marco de refrecia de la tierra es: 10.1010101010101 años


#### **(b)** como lo percibiría un pasajero a bordo de la nave. 

Por otro lado, el pasajero experimenta un tiempo propio $\tau$ debido a que su marco de referencia es inercial; por lo tanto: $$t = \gamma \tau$$

Donde $\gamma$ es el factor de Lorentz: $$\gamma = \sqrt{1-(\frac{v}{c})^{2}}$$

Despejando a $\tau$: $$\tau = t \sqrt{1-(\frac{v}{c})^{2}}$$

In [18]:
#EL programa seria:
#Como el tiempo esta en años luz se cancela la c dentro de la raiz
τ = t*(mt.sqrt(1-(v**2)))
print(f"El tiempo que tarda desde el marco de referencia de la nave es: {τ} años")

El tiempo que tarda desde el marco de referencia de la nave es: 1.424922826228878 años


#### Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

Por lo tanto, si el planeta está a 10 años luz y su velocidad es de $0.99c$, el tiempo que tarda en llegar desde el marco de referencia de la Tierra sería: $$t = 10.101 años$$

Asi, el tiempo que tarda desde el marco de referencia de la nave es: $$ \tau = 1.424 años$$

### **4. Paso de potencial cuántico.** 
#### Un conocido problema de mecánica cuántica involucra una partícula de masa $m$ que encuentra un paso de potencial unidimensional, como este:

<center><img src="qstep2.png" width="50%" class="center"/></center>

#### La partícula con energía cinética inicial $E$ y vector de onda $k_1=\sqrt{2mE}/\hbar$ entra por la izquierda y encuentra un salto repentino en la energía potencial de altura $V$ en la posición $x=0$. Resolviendo la ecuación de Schrödinger, se puede demostrar que cuando $E>V$ la partícula puede 

#### (a) **pasar el escalón**, en cuyo caso tiene una energía cinética menor a $E-V$ en el otro lado y un vector de onda correspondientemente menor a $k_2=\sqrt{2m(E-V)}/\hbar$, \; o bien (b) **reflejarse**, manteniendo toda su energía cinética y un vector de onda sin cambios, pero moviéndose en la dirección opuesta. Las probabilidades $T$ y $R$ de transmisión y reflexión están dadas por:

<h3 align="center">$T = {4k_1k_2\over(k_1+k_2)^2}\,,\qquad$
$R = \biggl( {k_1-k_2\over k_1+k_2} \biggr)^2.$</h3>

#### Supongamos que tenemos una partícula con una masa igual a la masa del electrón. $m=9.11\times10^{-31}\,$kg y energía de $10\,$eV, al encontrar un escalón de potencial de altura $9\,$eV. Escribe un programa en \texttt{Python} para calcular e imprimir las probabilidades de transmisión y reflexión utilizando las fórmulas anteriores.


In [11]:
#Primero definimos el valor de cada variable que va a ser constante
#Colocamos el valor de h barra en eV
ℏ = 6.58211e-16
#La masa
m = 9.11e-31
#La energia en eV
E = 10
#El potencial en eV
V = 9

In [12]:
#Ahora calculamos el valor de k1 y k2
k1 = (mt.sqrt(2*m*E))/ℏ
k2 = (mt.sqrt(2*m*(E-V)))/ℏ

In [19]:
#Por ultimo le pedimos que nos pida si queremos ver las probabilidades
while (True):
    p = str(input("Coloque ':D' si gusta la probabilidad de transmitancia y reflectancia: "))
    if (p == ':D'):
        T = (4*k1*k2)/((k1+k2)**2)
        R = ((k1-k2)/(k1+k2))**2
        print(f"La probabilidad de transmitancia es: {T}")
        print(f"La probabilidad de reflectancia es: {R}")
        break
    else:
        print("No quiere las probabilidades??")
        print("Ta bien le doy otra oportunidad")

Coloque ':D' si gusta la probabilidad de transmitancia y reflectancia: no
No quiere las probabilidades??
Ta bien le doy otra oportunidad
Coloque ':D' si gusta la probabilidad de transmitancia y reflectancia: :D
La probabilidad de transmitancia es: 0.7301261363877617
La probabilidad de reflectancia es: 0.2698738636122385
